crates/relayer/src/chain/ckb4ibc/monitor.rs

```diff
@@ -0,0 +1,427 @@
+use ibc_relayer_types::timestamp::Timestamp;
+use tokio::runtime::Runtime as TokioRuntime;
+
+use crate::chain::ckb::prelude::CkbReader;
+use crate::chain::ckb::rpc_client::RpcClient;
+use crate::chain::ckb4ibc::extractor::{
+    extract_channel_end_from_tx, extract_ibc_connections_from_tx, extract_ibc_packet_from_tx,
+};
+use crate::chain::tracking::TrackingId;
+use crate::config::ckb4ibc::ChainConfig;
+use crate::event::bus::EventBus;
+use crate::event::monitor::{Error, EventBatch, MonitorCmd, Next, Result, TxMonitorCmd};
+use crate::event::IbcEventWithHeight;
+
+use super::cache_set::CacheSet;
+
+// todo add cell emitter here
+pub struct Ckb4IbcEventMonitor {
+    rt: Arc<TokioRuntime>,
+    rpc_client: Arc<RpcClient>,
+    rx_cmd: Receiver<MonitorCmd>,
+    event_bus: EventBus<Arc<Result<EventBatch>>>,
+    config: ChainConfig,
+    cache_set: RwLock<CacheSet<H256>>,
+}
+
+impl Ckb4IbcEventMonitor {
+    pub fn new(
+        rt: Arc<TokioRuntime>,
+        rpc_client: Arc<RpcClient>,
+        config: ChainConfig,
+    ) -> (Self, TxMonitorCmd) {
+        let (tx_cmd, rx_cmd) = crossbeam_channel::unbounded();
+        let monitor = Ckb4IbcEventMonitor {
+            rt,
+            rpc_client,
+            rx_cmd,
+            event_bus: EventBus::default(),
+            config,
+            cache_set: RwLock::new(CacheSet::new(512)),
+        };
+        (monitor, TxMonitorCmd::new(tx_cmd))
+    }
+
+    pub fn run(mut self) {
+        let rt = self.rt.clone();
+        loop {
+            let result = rt.block_on(self.run_once());
+            match result {
+                Next::Continue => continue,
+                Next::Abort => break,
+            }
+        }
+    }
+    async fn run_once(&mut self) -> Next {
+        if let Ok(cmd) = self.rx_cmd.try_recv() {
+            match cmd {
+                MonitorCmd::Shutdown => return Next::Abort,
+                MonitorCmd::Subscribe(tx) => tx.send(self.event_bus.subscribe()).unwrap(),
+            }
+        }
+        let result = async {
+            tokio::select! {
+                Ok(batch) = self.fetch_channel_events() => {
+                    batch
+                },
+                Ok(batch) = self.fetch_connection_events() => {
+                    batch
+                },
+                Ok(batch) = self.fetch_packet_events() => {
+                    batch
+                }
+            }
+        }
+        .await;
+
+        self.process_batch(result);
+        Next::Continue
+    }
+
+    async fn fetch_connection_events(&self) -> Result<EventBatch> {
+        let key = SearchKey {
+            script: Script::new_builder()
+                .code_hash(self.config.connection_code_hash.pack())
+                .args(self.config.connection_args.pack())
+                .build()
+                .into(),
+            script_type: ScriptType::Lock,
+            filter: None,
+            with_data: None,
+            group_by_transaction: None,
+        };
```

Comment:
I highly recommend you wrap these useful code snippets as a utility function to use, it's better for the code reading and comprehension